### PR TITLE
- Renamed variables, removed unused logic, and commented code.

### PR DESCRIPTION
**Pull request by danec020**
Added comments and renamed variable to the previous changes to make it more understandable. I found using the longer variable names made it much easier to follow the flow of the logic due to the finicky nature of dealing with children's children of parent's children ect...

This part was removed from IF statement to see how many sized_zones are children of the tab's parent but could be used later if you need to propagate other type of widgets up to the parent of the removed zone?

```
//If nothing is found then the docking_zone can be removed. NOTE: This doesn't mean
//there aren't children inside that are of a different type than sized_zones.
if child_zones_count.len() == 0 {
    if let Ok(tz_parents_parent) = q_parent.get(tab_zone_parent_id) {
        let tz_parents_parent = tz_parents_parent.get();
        //Find out what index the tab's parent is so we can remove it (Contains all Tab elements)
        let removal_index = q_children
            .get(tz_parents_parent)
            .unwrap()
            .iter()
            .position(|child| *child == tab_zone_parent_id)
            .unwrap();

        Now that we have the index that is going to be removed, move it's children up.
        Earlier we only looked for children of sized_zones under this parent, not other types.
        for child in tz_parent_children {
            //Ignore the tab_zone that is going to be removed, it is no longer needed.
            if *child == tab_zone.zone { continue; }

            //Put the children at the index this zone use to be in
            //*NOTE: q_sized_zones in this example has already be searched on line 84 and
            //we have determined child cound is 0, therefore this can never happen. But
            //this example may be useful if you were looking for something OTHER than a
            //zone to throw these children in? 
            if q_sized_zones.get(*child).is_ok() {
                commands
                    .entity(tz_parents_parent)
                    .insert_children(removal_index, &[*child]);
            }
        }
    }
    commands.entity(tab_zone_parent_id).despawn_recursive();
    despawn_zone = false;
}
```

### CODE REVIEW

File: src/widgets/docking_zone.rs [hunk 1]
```diff
@@ -46,77 +46,58 @@ impl Plugin for DockingZonePlugin {
...
+            //By keeping count of the children in a vector we also know how many zones are availble as well.
+            let mut child_zones_count: Vec<usize> = vec![];
+            if let Ok(tz_parent_children) = q_children.get(tab_zone_parent_id) {
+                for child in tz_parent_children {
+                    //Ignore the zone we wish to remove, it is empty
+                    if *child == tab_zone.zone { continue; }
+                    //Check if this child is a zone that may need to be later propagated
+                    //upwards if this zone is removed. Store how many children it has.
+                    if q_sized_zones.get(*child).is_ok() {
+                        child_zones_count.push(tz_parent_children.iter()
```
Comment (eidloi) [inline src/widgets/docking_zone.rs:84]: won't this count the number of siblings instead of the number of child elements of the zone?

File: src/widgets/docking_zone.rs [hunk 2]
```diff
@@ -46,77 +46,58 @@ impl Plugin for DockingZonePlugin {
...
+                //If nothing is found then the docking_zone can be removed. NOTE: This doesn't mean
+                //there aren't children inside that are of a different type than sized_zones.
+                if child_zones_count.len() == 0 {
+                    commands.entity(tab_zone_parent_id).despawn_recursive();
                     despawn_zone = false;
                 }
             }
         }
-
+        //We didn't have to dispawn the parent of this tab, so despawn this zone and reset the parent.
```
Comment (eidloi) [inline src/widgets/docking_zone.rs:97]: this comment is weird. we did have to despawn the containing zone, which is the parent of the tab! It would make more sense to say: "We don't need to keep the zone, there are no sibling zones"

**Review by eidloi — CHANGES_REQUESTED** [review]

File: src/widgets/docking_zone.rs [hunk 3]
```diff
@@ -46,77 +46,58 @@ impl Plugin for DockingZonePlugin {
 #[derive(SystemSet, Clone, Eq, Debug, Hash, PartialEq)]
 pub struct DockingZoneUpdate;
 
-// TODO: Fix lingering DockingZoneSplitContainer after their sole docking zone child's content is removed
```
Comment (eidloi) [inline src/widgets/docking_zone.rs:49]: Changes here won't fix it actually. We'll need a separate system probably. The issue is that wrappers of sized zones remain. I should probably draw a picture of this issue specifically

Comment (eidloi) [reply src/widgets/docking_zone.rs:49]: zone split example
the use case I refer to in the TODO can be observed at the rightmost state: if you remove the blue and orange zones (because their tab containers got emptied) two extra zones remain that wrap the remaining three rows, which are unnecessary (and bloat the borders)

**Comment by danec020** [comment]
Closed this pull request due to it being out of date. The issues @eidloi has mentioned needed to be addressed as this PR did not handle it properly. A new PR will be submitted with the correct changes.

**Comment by danec020** [comment]
The new changes submitted should address all the concerns mentioned by @eidloi.

Now all remaining children who are not related to the requested zone will shift up if necessary. This creates it's own issues since the UI layout will possibly change. There is a sliced_test.rs example that I have added to my fork that can be used to test this edge case. The result will be a sliced zone that has a sliced zone child containing only a label. When this happens the child will be removed and the label will shift up to the parent sliced zone. This seems like the logical solution but may need to be reconsidered?

Also with these changes, removed zones will grab their resizehandle children so they don't clutter up the hierarchy after the zone is despawned.

File: src/widgets/docking_zone.rs (hunk 1 above)
Comment (danec020) [reply src/widgets/docking_zone.rs]: This has been updated.

We now get all the children regardless if they are of sized/sliced zone as well as the amount of child zones other than the requested one. Lastly we have a bool set that lets us know if any of these other zones are of type split zone. This is used when only 1 child zone is found, we can then see if it was a split zone.

File: src/widgets/docking_zone.rs (hunk 2 above)
Comment (danec020) [reply src/widgets/docking_zone.rs]: I removed the confusing comment.

File: src/widgets/docking_zone.rs (hunk 3 above)
Comment (danec020) [reply src/widgets/docking_zone.rs:49]: I believe this issue is resolved now, please confirm.

**Review by eidloi — COMMENTED** [review]
Ahoy! I know you worked a lot on this, but I'll reject this PR. Apologies.

This is probably the most complex widget in the library so far with some very special needs and a lot of edge cases. I'll do a fix in a couple of days (once I am done with the theming) and will create a PR. We can do a review together with screen-share so I can better explain all the details.

And thank for pointing out some more edge cases!

File: src/widgets/docking_zone.rs [hunk 4]
```diff
@@ -46,77 +46,98 @@ impl Plugin for DockingZonePlugin {
...
+                    children_to_shift.push(*child);
+
+                    //Check if this child is a sized_zone that may need to be later propagated
+                    //upwards if the requested zone is removed or repositioned.
+                    //NOTE: A SplitZone is always paired with a SizedZone
+                    if q_sized_zones.get(*child).is_ok() {
+                        remaining_zones += 1;
+
+                        //Check to see if the child is a split_zone to handle edge case for nested split_zones
+                        if !remaining_has_split
```
[the remark this answers is not in the document]
Comment (danec020) [reply src/widgets/docking_zone.rs:99]: If it was previously set it could be set to false if their was another zone as a child. But then we wouldn't need this flag as the zone child count != 1, so you are correct. Thanks for the catch.

File: src/widgets/docking_zone.rs [hunk 5]
```diff
@@ -46,77 +46,98 @@ impl Plugin for DockingZonePlugin {
...
+                        {
```
[the remark this answers is not in the document]
Comment (danec020) [reply src/widgets/docking_zone.rs:100]: Fixed

File: src/widgets/docking_zone.rs [hunk 6]
```diff
@@ -46,77 +46,98 @@ impl Plugin for DockingZonePlugin {
...
-
-                    if q_sized_zone.get(*child).is_ok() {
-                        zone_child_count.push(children.iter()
-                        .filter(|child| q_sized_zone.get(**child).is_ok())
-                        .count());
+        let tab_zone_parent_id = tab_zone_parent.get();
+        if let Ok(_) = q_split_zones.get(tab_zone_parent_id) {
+            //Store the children that may need to be shifted if this becomes a nested split zone.
+            //Set a bool to let the system know that the child is of type split zone as well.
+            let mut children_to_shift: Vec<Entity> = vec![];
```
[the remark this answers is not in the document]
Comment (danec020) [reply src/widgets/docking_zone.rs:78]: The TabContainer's DockingZone is the item marked with RemoveEmptyDockingZone, not its parent DockingZoneSplitContainer. We are only shifting the children who aren't related to the RemoveEmptyDockingZone, otherwise you will end up with a DockingZoneSplitContainer containing a single another DockingZoneSplitContainer. 

From my understanding, this is not ever a desired outcome? When this happens it is no longer to ever dock a zone in the top level parent unless you close the obstructing zone. The steps this change does removes the extra DockingZoneSplitContainer.

**Comment by danec020** [comment]
No problem. Sorry to have wasted your time.

**Comment by eidloi** [comment]
> No problem. Sorry to have wasted your time.

You didn't waste it mate! Your research is valuable, and will contribute to the overall fix!
